sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -573,30 +573,79 @@ object GpuCast extends Arm {
       inclusiveMax: Boolean = true,
       errorMsg:String = GpuCast.OVERFLOW_MESSAGE): Unit = {
 
-    def throwIfAny(cv: ColumnView): Unit = {
-      withResource(cv) { cv =>
-        withResource(cv.any()) { isAny =>
-          if (isAny.isValid && isAny.getBoolean) {
+    def throwAnyNan(): Unit = {
+      withResource(values.isNan()) { valuesIsNan =>
+        withResource(valuesIsNan.any()) { anyNan =>
+          if (anyNan.isValid && anyNan.getBoolean) {
             throw RapidsErrorUtils.arithmeticOverflowError(errorMsg)
           }
         }
       }
     }
 
-    withResource(minValue) { minValue =>
-      throwIfAny(if (inclusiveMin) {
-        values.lessThan(minValue)
-      } else {
-        values.lessOrEqualTo(minValue)
-      })
+    def throwOutOfRange[T](minInput: T, minValue: T,
+                           maxInput: T, maxValue: T)
+                          (implicit ord: Ordering[T]): Unit = {
+      if (inclusiveMin && ord.compare(minInput, minValue) < 0 ||
+          !inclusiveMin && ord.compare(minInput, minValue) <= 0 ||
+          inclusiveMax && ord.compare(maxInput, maxValue) > 0 ||
+          !inclusiveMax && ord.compare(maxInput, maxValue) >= 0) {
+        throw RapidsErrorUtils.arithmeticOverflowError(errorMsg)
+      }
     }
 
-    withResource(maxValue) { maxValue =>
-      throwIfAny(if (inclusiveMax) {
-        values.greaterThan(maxValue)
-      } else {
-        values.greaterOrEqualTo(maxValue)
-      })
+    withResource(minValue) { minValue =>
+      withResource(maxValue) { maxValue =>
+        withResource(values.min()) { minInput =>
+          withResource(values.max()) { maxInput =>
+            values.getType match {
+              case DType.FLOAT64 =>
+                throwAnyNan()
+                def getDoubleValue(s: Scalar): Double = s.getType match {
+                  case DType.INT8 | DType.UINT8 => s.getByte.toDouble
+                  case DType.INT16 | DType.UINT16 => s.getShort.toDouble
+                  case DType.INT32 | DType.UINT32 => s.getInt.toDouble
+                  case DType.INT64 | DType.UINT64 => s.getLong.toDouble
+                  case _ => s.getDouble
+                }
```

Comment:
```suggestion
                def getDoubleValue(s: Scalar): Double = s.getType match {
                  case DType.FLOAT64 => s.getDouble
                  case _ => throw IllegalArgumentException(...)
                }
```

Reply:
Thank you for your review! When I make this change, I failed 5 unit tests, all throwing the exception in line 606. (The test cases are "Write doubles to long/int/short/byte (values within range)" from AnsiCastOpSuite)

Reply:
Can you point out what tests failed?

Reply:
Ah I know why they failed. That's because:
```
case (FloatType | DoubleType, IntegerType) if ansiMode =>
        assertValuesInRange(input, Scalar.fromInt(Int.MinValue),
          Scalar.fromInt(Int.MaxValue))
        input.castTo(GpuColumnVector.getNonNestedRapidsType(toDataType))
```

Reply:
Alright, then how about this:
```
.....
                  case DType.FLOAT32 => s.getFloat.toDouble
                  case DType.FLOAT64 => s.getDouble
                  case _ => throw ...
                }
```

That's because we should also handle the cases of float/double correctly.

Reply:
I am calling this helper function on the `min/maxValue` passed into `assertValuesInRange` as `Scalar.fromInt(Int.MinValue)` and `Scalar.fromInt(Int.MaxValue)` in the previous example. Therefore I am casing on the byte/short/int/long type. I ran into the same failures with the above code.

Reply:
The code snippet above is to add to your function, not to replace it.

Reply:
You can do better by using just one function (below) for both float and double:
```
  def getValue[T](s: Scalar): T = (s.getType match {
                  case DType.INT8 | DType.UINT8 => s.getByte
                  case DType.INT16 | DType.UINT16 => s.getShort
                  case DType.INT32 | DType.UINT32 => s.getInt
                  case DType.INT64 | DType.UINT64 => s.getLong
                  case DType.FLOAT32 => s.getFloat
                  case DType.FLOAT64 => s.getDouble
                  case _ => s.getFloat // <= not sure if this line is compiled---remove this if not
).asInstanceOf[T]
```

Reply:
I was trying the above implementation but I ran into issues like `java.lang.Long cannot be cast to java.lang.Float`. It seems like we have to call the `toFloat/Double` function explicitly. I am now working on changing the signature of assertValuesInRange and see if that could help us remove these redundant matchings. Thanks!